### ISSUE TITLE
v0.11: high CPU/net on client after broker reboot

### ISSUE THREAD

**Issue by zarnovican** [edenhill/librdkafka#1389]
Description
===========

Several minutes after broker reboot, producer client will enter a state with high CPU usage, low msg rate and high network traffic throughput.

Details:
Our test setup has 3-node Kafka+Zookeeper cluster on AWS with 0.11.0.0, one producer and one consumer. Both are written in Go, using librdkafka 0.11.0. Producer is generating ~300msg/s, which is ~120kB/s into a topic with 12 partitions and replication factor 2.

Right after broker reboot, it appears that the msg rate is maintained at the same level 300msg/s. Several minutes later (3-4min), msg rate drops to approx 2/3 (200msg/s). At the same time, producer's CPU hits the ceiling and network throughput rises to 40MB/s (that *340x* of normal traffic!!). This period lasts from couple of seconds to several minutes. At the end, producer seems to delived the missing 1/3 of messages after all and everything returns to normal.

Network and CPU from host running producer/consumer:
![image](https://user-images.githubusercontent.com/781024/29563319-46729662-873d-11e7-803d-2e55855a115e.png)
![image](https://user-images.githubusercontent.com/781024/29563325-4b34aa0a-873d-11e7-9fbb-624218455c67.png)

I have managed to capture couple of seconds of traffic during this abnormal period. It seems that producer is repeatedly trying to send large chunks of data to non-leader broker. The size of the chunk is growing over time and there may be several of them (one per broken partition?). At our rate ~120kB/s, we should have ~10kB/s per partition (12), so 300kB would be the size of 30s of data.

![image](https://user-images.githubusercontent.com/781024/29563448-ec037d9e-873d-11e7-863f-1164ca0cfe85.png)

We have seen the same high CPU issue on consumer side as well. There were also cases when the consumer haven't recovered and stayed at <300 msg/s until restarted.

I have to admit, that we don't have much prior exp with Kafka, and the "latest stable" 0.11 is the first version we tried.

How to reproduce
================

I don't have a simple, consistent reproducer. I would only _assume_ that on any 3-node Kafka cluster with `librdclient` of the same version, you should be able to reproduce it as well.

We might be able to provide local (Docker?) setup with similar client in the future. Time permiting.

1. start 3-node Kafka+Zookeeper cluster on 0.11.0.0
1. create replicated/paritioned topic
1. start producer/consumer on that topic with reasonable msg rate (300msg/s in our case)
1. reboot one of the kafka brokers
1. observe CPU/net resource usage and msg rate on clients for several minutes following reboot.

Checklist
=========
Please provide the following information:

 - [x] librdkafka version (release number or git tag): 0.11.0
 - [x] Apache Kafka version: kafka_2.11-0.11.0.0
 - [x] librdkafka client configuration: most of it is on defaults AFAIK
 - [x] Operating system: Ubuntu 16.04
 - [ ] Using the legacy Consumer
 - [ ] Using the high-level KafkaConsumer
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [x] Provide broker log excerpts
```
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherManager [ReplicaFetcherManager on broker 1] Removed fetcher for partitions __consumer_offsets-19
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  cluster.Partition Partition [__consumer_offsets,19] on broker 1: __consumer_offsets-19 starts at Leader Epoch 46 from offset 0. Previous Leader Epoch was: 45
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  group.GroupMetadataManager [Group Metadata Manager on Broker 1]: Scheduling loading of offsets and group metadata from __consumer_offsets-19
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherThread [ReplicaFetcherThread-0-3]: Error for partition [data.messages,2] to broker 3:org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherManager [ReplicaFetcherManager on broker 1] Removed fetcher for partitions data.messages-2
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  cluster.Partition Partition [data.messages,2] on broker 1: data.messages-2 starts at Leader Epoch 28 from offset 816281. Previous Leader Epoch was: 27
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherThread [ReplicaFetcherThread-0-2]: Error for partition [data.streams,7] to broker 2:org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  group.GroupMetadataManager [Group Metadata Manager on Broker 1]: Finished loading offsets and group metadata from __consumer_offsets-19 in 7 milliseconds.
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherThread [ReplicaFetcherThread-0-3]: Error for partition [__consumer_offsets,10] to broker 3:org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherManager [ReplicaFetcherManager on broker 1] Removed fetcher for partitions data.streams-7
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  cluster.Partition Partition [data.streams,7] on broker 1: data.streams-7 starts at Leader Epoch 12 from offset 32. Previous Leader Epoch was: 11
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherThread [ReplicaFetcherThread-0-3]: Error for partition [__consumer_offsets,40] to broker 3:org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  server.ReplicaFetcherManager [ReplicaFetcherManager on broker 1] Removed fetcher for partitions __consumer_offsets-10
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  cluster.Partition Partition [__consumer_offsets,10] on broker 1: __consumer_offsets-10 starts at Leader Epoch 43 from offset 0. Previous Leader Epoch was: 42
Aug 22 10:45:31 ip-10-99-139-184 kafka[1130]:  group.GroupMetadataManager [Group Metadata Manager on Broker 1]: Scheduling loading of offsets and group metadata from __consumer_offsets-10
```
 - [x] Critical issue: It is not a _huge_ issue that producer is degraded for several minutes. More worrying for me is that it will consumer all CPU/bandwidth for that host.

**Comment by zarnovican** [comment]
## Further Info

I have repeated the test with `debug=broker,topic` and `tcpdump` sniffing Kafka traffic up to the point when the period with high network traffic starts.

It confirms the previous assumption that "Not leader for.." responses are received (and only) for partitions originally handled by the rebooted broker.

At .114 broker 2 will start responding with "Not Leader.."
```
%7|1503493624.114|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.115|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [6]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493624.115|FASTQUERY|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: Starting fast leader query
%7|1503493624.116|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.116|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [6]: leader unavailable: produce: Broker: Not leader for partition
```

At .430 broker 2 closed the TCP session. This is immediately followed by Metadata request from client and all four partitions migrate to remaining brokers.
```
%7|1503493624.405|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.430|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [3]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493624.430|BROKERFAIL|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: failed: err: Local: Broker transport failure: (errno: Resource temporarily unavailable)
%3|1503493624.430|FAIL|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Receive failed: Disconnected
%3|1503493624.430|ERROR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Receive failed: Disconnected
%7|1503493624.430|STATE|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Broker changed state UP -> DOWN
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 8 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 2 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 11 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 5 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 4 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 7 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 1 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 10 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 9 Leader 3
%7|1503493624.431|TOPICUPD|mqtt_reader.1#producer-1| [thrd:main]: Topic data.messages [9] migrated from broker 2 to 3
%7|1503493624.431|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [9]: delegate to broker ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 (rktp 0x2e18cc0, term 0, ref 4, remove 0)
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [9]: broker ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 no longer leader
```
Client keeps trying to reconnect to downed broker:
```
%7|1503493627.577|FAIL|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Connect to ipv4#10.192.128.217:9092 failed: Connection refused
%7|1503493627.577|STATE|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Broker changed state CONNECT -> DOWN
%7|1503493628.576|CONNECT|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: broker in state DOWN connecting
%7|1503493628.578|CONNECT|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Connecting to ipv4#10.192.128.217:9092 (plaintext) with socket 18
%7|1503493628.578|STATE|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Broker changed state DOWN -> CONNECT
```
For a minute or so, all traffic is sent to the two remaining brokers.

Then "Not Leader" responses start..
```
%7|1503493693.024|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493693.024|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: data.messages [6]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493693.024|FASTQUERY|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: Starting fast leader query
%7|1503493693.025|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493693.026|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: data.messages [6]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493693.027|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
```
At the same time, Kafka log indicates that the partition leader has changed. Rebooted broker became leader for that partition again.
```
Aug 23 13:08:13 ip-10-99-139-184 kafka[1130]:  controller.KafkaController [Controller 1]: Partition [data.messages,6] completed preferred replica leader election. New leader is 2
```

Client keeps pushing to non-leaders for 174s, until it will make Metadata request again.
```
%7|1503493867.074|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: ec2-54-78-20-139.eu-west-1.compute.amazonaw
s.com:9092/1: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493867.076|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: data.messages [6]: leader unavailable: pr
oduce: Broker: Not leader for partition
%7|1503493867.082|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 8 Leader 1
%7|1503493867.082|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 2 Leader 1
%7|1503493867.082|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 11 Leader 1
```

This 3min period is when client generates such large network traffic and CPU load.

### Observations

During broker failure, `librdkafka` updates its metadata cache quickly. There is only 300ms between first "Not leader" and Metadata request call. This request seems to be triggered by loss of TCP connection rather than "Not leader .." response.

During broker recovery nothing seems to be triggering Metadata cache update and client keeps sending data to non-leader brokers. Up until the metadata is updated by other means (periodic refresh?).

High CPU load and network traffic is (most likely) caused by client having stale metadata cache for a prolonged period of time and at the same time aggressively retrying to re-send data to wrong brokers.

There is debug log entry `Starting fast leader query` at the right place where I would expect client to refresh metadata info. However, I don't see any packet in tcpdump that would correspond to that debug message. Is this a client bug or a feature that must be explicitly enabled ?

Or is it library user's responsibility to make a call to refresh metadata cache explicitly in one of the callbacks (on_deliver, on_error..) ?

**Comment by edenhill** [comment]
Thanks for a great report!

I'm missing some information from the logs, in particular the METADATA reply seen in your second paste does not contain the topic pre-amble nor all partitions for the topic. Can you send me the full client log file if possible?

**Comment by zarnovican** [comment]
Here are all METADATA messages around that second, plus some context.
```
%7|1503493624.404|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.404|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [0]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493624.404|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.405|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [9]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493624.405|REQERR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: ProduceRequest failed: Broker: Not leader for partition: explicit actions 0x0
%7|1503493624.430|LEADERUA|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: data.messages [3]: leader unavailable: produce: Broker: Not leader for partition
%7|1503493624.430|BROKERFAIL|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: failed: err: Local: Broker transport failure: (errno: Resource temporarily unavailable)
%3|1503493624.430|FAIL|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Receive failed: Disconnected
%3|1503493624.430|ERROR|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Receive failed: Disconnected
%7|1503493624.430|STATE|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Broker changed state UP -> DOWN
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 8 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 2 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 11 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 5 Leader 1
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 4 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 7 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 1 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 10 Leader 3
%7|1503493624.431|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 9 Leader 3
%7|1503493624.431|TOPICUPD|mqtt_reader.1#producer-1| [thrd:main]: Topic data.messages [9] migrated from broker 2 to 3
%7|1503493624.431|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [9]: delegate to broker ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 (rktp 0x2e18cc0, term 0, ref 4, remove 0)
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [9]: broker ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 no longer leader
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [9]: broker ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 is now leader for partition with 0 messages (0 bytes) queued
%7|1503493624.432|BRKMIGR|mqtt_reader.1#producer-1| [thrd:main]: Migrating topic data.messages [9] 0x2e18cc0 from ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 to ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 (sending PARTITION_LEAVE to ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2)
%7|1503493624.432|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 3 Leader 3
%7|1503493624.432|TOPICUPD|mqtt_reader.1#producer-1| [thrd:main]: Topic data.messages [3] migrated from broker 2 to 3
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [3]: delegate to broker ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 (rktp 0x2e17100, term 0, ref 3, remove 0)
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [3]: broker ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 no longer leader
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [3]: broker ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 is now leader for partition with 4 messages (666 bytes) queued
%7|1503493624.432|BRKMIGR|mqtt_reader.1#producer-1| [thrd:main]: Migrating topic data.messages [3] 0x2e17100 from ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 to ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3 (sending PARTITION_LEAVE to ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2)
%7|1503493624.432|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 6 Leader 1
%7|1503493624.432|TOPICUPD|mqtt_reader.1#producer-1| [thrd:main]: Topic data.messages [6] migrated from broker 2 to 1
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [6]: delegate to broker ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 (rktp 0x2e17ee0, term 0, ref 4, remove 0)
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [6]: broker ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 no longer leader
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [6]: broker ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 is now leader for partition with 0 messages (0 bytes) queued
%7|1503493624.432|BRKMIGR|mqtt_reader.1#producer-1| [thrd:main]: Migrating topic data.messages [6] 0x2e17ee0 from ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 to ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 (sending PARTITION_LEAVE to ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2)
%7|1503493624.432|METADATA|mqtt_reader.1#producer-1| [thrd:main]:   Topic data.messages partition 0 Leader 1
%7|1503493624.432|TOPICUPD|mqtt_reader.1#producer-1| [thrd:main]: Topic data.messages [0] migrated from broker 2 to 1
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [0]: delegate to broker ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 (rktp 0x2e16320, term 0, ref 4, remove 0)
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [0]: broker ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 no longer leader
%7|1503493624.432|BRKDELGT|mqtt_reader.1#producer-1| [thrd:main]: data.messages [0]: broker ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 is now leader for partition with 0 messages (0 bytes) queued
%7|1503493624.432|BRKMIGR|mqtt_reader.1#producer-1| [thrd:main]: Migrating topic data.messages [0] 0x2e16320 from ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2 to ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1 (sending PARTITION_LEAVE to ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2)
%7|1503493624.432|PARTCNT|mqtt_reader.1#producer-1| [thrd:main]: Partitioning 0 unassigned messages in topic data.messages to 12 partitions
%7|1503493624.432|UAS|mqtt_reader.1#producer-1| [thrd:main]: 0/0 messages were partitioned in topic data.messages
%7|1503493624.432|METADATA|mqtt_reader.1#producer-1| [thrd:main]: kafka.nrgmntr.com:9092/bootstrap: 1/1 requested topic(s) seen in metadata
%7|1503493624.531|CONNECT|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: broker in state DOWN connecting
%7|1503493624.532|CONNECT|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Connecting to ipv4#10.192.128.217:9092 (plaintext) with socket 18
%7|1503493624.532|STATE|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Broker changed state DOWN -> CONNECT
%7|1503493624.533|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Topic data.messages [9]: leaving broker (0 messages in xmitq, next leader ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3, rktp 0x2e18cc0)
%7|1503493624.533|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3]: ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3: Topic data.messages [9]: joining broker (rktp 0x2e18cc0)
%7|1503493624.533|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Topic data.messages [3]: leaving broker (0 messages in xmitq, next leader ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3, rktp 0x2e17100)
%7|1503493624.533|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3]: ec2-54-216-54-205.eu-west-1.compute.amazonaws.com:9092/3: Topic data.messages [3]: joining broker (rktp 0x2e17100)
%7|1503493624.533|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Topic data.messages [6]: leaving broker (0 messages in xmitq, next leader ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1, rktp 0x2e17ee0)
%7|1503493624.534|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1]: ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1: Topic data.messages [6]: joining broker (rktp 0x2e17ee0)
%7|1503493624.534|TOPBRK|mqtt_reader.1#producer-1| [thrd:ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2]: ec2-54-216-200-214.eu-west-1.compute.amazonaws.com:9092/2: Topic data.messages [0]: leaving broker (0 messages in xmitq, next leader ec2-54-78-20-139.eu-west-1.compute.amazonaws.com:9092/1, rktp 0x2e16320)
```

**Comment by edenhill** [comment]
Thank you, I'd still like to get my hands on the full log to understand what is and isnt going on.

**Comment by zarnovican** [comment]
Here is the requested log (14MB) in full..
https://github.com/zarnovican/kafka_test/raw/master/mqtt_reader.log.gz

**Comment by edenhill** [comment]
While scanning through the logs I'm a bit concerned about the number of ProduceRequests, I can't tell for sure what is going on but I get the feeling that there are a huge number of outstanding ProduceRequests in-flight when the leader change occurs, and the number of ProduceRequests may be due to a too low `linger.ms` setting, causing batches to not really accumulate before sending.
Not saying this is the core issue, but I suggest you try increasing `linger.ms` to some value that makes sense for your throughput, say 50-500ms, to make sure each ProduceRequest accumulates a larger number of messages in a batch.

**Comment by tchap** [comment]
@edenhill We will try to tune it somehow, but it doesn't seem to be to core issue, as you mentioned.

**Comment by edenhill** [comment]
I'll try to reproduce this to see what's going on.

**Comment by zarnovican** [comment]
I have debugged the library code by adding some extra logging and came to the same root-cause and fix as for #1313 (commit cd4b629eac507afea8e837a0e2cb18392effcc38). I have re-tested the issue from `master` branch and it doesn't appear there anymore. Metadata on failure and recovery converge <1s for both producer and consumer.

Quoting comment from linked issue:
> "Starting fast leader query" appear per second，but it doesn't trigger the metedata request
> the last metadata request is triggered because periodic refresh, not the fast leader query

This pretty much re-iterates my own observations of 0.11.0 behavior.

I'm closing this as a duplicate of #1313.
